lib/serializer/cbor/aws_iot_serializer_tinycbor_encoder.c

```diff
@@ -0,0 +1,297 @@
+static AwsIotSerializerError_t _append( AwsIotSerializerEncoderObject_t * pEncoderObject,
+                                        AwsIotSerializerScalarData_t scalarData);
+static AwsIotSerializerError_t _appendKeyValue( AwsIotSerializerEncoderObject_t * pEncoderObject,
+                                                const char * pKey,
+                                                AwsIotSerializerScalarData_t scalarData);
+
+
+AwsIotSerializerEncodeInterface_t _AwsIotSerializerCborEncoder =
+{
+    .getEncodedSize           = _getEncodedSize,
+    .getExtraBufferSizeNeeded = _getExtraBufferSizeNeeded,
+    .init                     = _init,
+    .destroy                  = _destroy,
+    .openContainer            = _openContainer,
+    .openContainerWithKey     = _openContainerWithKey,
+    .closeContainer           = _closeContainer,
+    .append                   = _append,
+    .appendKeyValue           = _appendKeyValue,
+};
+
+
+/*-----------------------------------------------------------*/
+
+static size_t _getEncodedSize( AwsIotSerializerEncoderObject_t * pEncoderObject,
+                               uint8_t * pDataBuffer )
+{
+    CborEncoder * pCborEncoder = _castEncoderObjectToCborEncoder( pEncoderObject );
+
+    return cbor_encoder_get_buffer_size( pCborEncoder, pDataBuffer );
+}
+
+/*-----------------------------------------------------------*/
+
+static size_t _getExtraBufferSizeNeeded( AwsIotSerializerEncoderObject_t * pEncoderObject )
+{
+    CborEncoder * pCborEncoder = _castEncoderObjectToCborEncoder( pEncoderObject );
+
+    return cbor_encoder_get_extra_bytes_needed( pCborEncoder );
+}
+
+/*-----------------------------------------------------------*/
+
+static AwsIotSerializerError_t _init( AwsIotSerializerEncoderObject_t * pEncoderObject,
+                                      uint8_t * pDataBuffer,
+                                      size_t maxSize )
+{
+    AwsIotSerializerError_t error = AWS_IOT_SERIALIZER_SUCCESS;
+    /* unuseful flags for tinycbor init */
+    int unusedCborFlags = 0;
+
+    CborEncoder * pCborEncoder = pvPortMalloc( sizeof( CborEncoder ) );
+
+    /* malloc succeeded */
+    if( pCborEncoder != NULL )
+    {
+        /* store the CborEncoder pointer to handle */
+        pEncoderObject->pHandle = pCborEncoder;
+        pEncoderObject->type = AWS_IOT_SERIALIZER_CONTAINER_STREAM;
+        cbor_encoder_init( pCborEncoder, pDataBuffer, maxSize, unusedCborFlags );
+    }
+    else /* malloc failed */
+    {
+        /* pEncoderObject is untouched. */
+
+        error = AWS_IOT_SERIALIZER_OUT_OF_MEMORY;
+    }
+
+    return error;
+}
+
+/*-----------------------------------------------------------*/
+
+static AwsIotSerializerError_t _destroy( AwsIotSerializerEncoderObject_t * pEncoderObject )
+{
+    CborEncoder * pCborEncoder = _castEncoderObjectToCborEncoder( pEncoderObject );
+
+    vPortFree( pCborEncoder );
+    _castEncoderObjectToCborEncoder( pEncoderObject ) = NULL;
```

Comment:
Does this line do anything?

Reply:
It sets the internal encoder object handle to NULL

Reply:
I get it now. It makes sense.